flake: Cloud events unit tests

Both `TestReconcile_CloudEvents` in `pkg/reconciler/pipelinerun/pipelinerun_test.go` with:
```
{Failed      pipelinerun_test.go:6018: config maps: [&ConfigMap{ObjectMeta:{config-defaults  knative-testing    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Data:map[string]string{default-cloud-events-sink: http://synk:8080,},BinaryData:map[string][]byte{},Immutable:nil,}]
          }
          }}
```

...and `TestReconcile_CloudEvents/Run_with_no_condition` in `pkg/reconciler/run/run_test.go` with:
```
{Failed      run_test.go:246: error in test Run with no condition: timed out waiting for 1 more events: []string{"(?s)dev.tekton.event.run.started.v1.*test-run"}
    run_test.go:255: error in test Run with no condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.started.v1\n  source: /apis///namespaces/foo//test-run\n  subject: test-run\n  id: 3d7cd7c6-777f-466d-9c0e-99bf34c99a3e\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n        \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"extraFields\": null\n      }\n    }\n  }\n"
INFO	TestReconcile_CloudEvents/Run_with_no_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
  specversion: 1.0
  type: dev.tekton.event.run.started.v1
  source: /apis///namespaces/foo//test-run
  subject: test-run
  id: 318d26ad-b6a3-4944-9ca4-47918e2b9028
  datacontenttype: application/json
Data,
  {
    "run": {
      "metadata": {
        "name": "test-run",
        "namespace": "foo",
        "resourceVersion": "00001",
        "creationTimestamp": null
      },
      "spec": {
        "serviceAccountName": ""
      },
      "status": {
        "extraFields": null
      }
    }
  }
 already sent}
```

...have been failing sporadically in CI lately.

/kind flake

Comparing a failed run of a test with a successful one:

Failed:
```
    logger.go:130: 2022-08-11T16:04:08.089Z     INFO    TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-11T16:04:08.090Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:67   Emitting cloudevent for test-run, condition:
 &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T16:04:08.092Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T16:04:08.170Z     INFO    TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-11T16:04:08.170Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      run/run.go:67   Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T16:04:08.172Z     DEBUG   TestReconcile_CloudEvents/Run_with_finished_true_condition      cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    run_test.go:255: error in test Run with finished true condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.successful.v1\n  source: /apis///namespaces/foo//test-run\n  subject: test-run\n  id: 697d6d54-5e82-4417-b1ff-27461844d2a0\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n        \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"conditions\": [\n          {\n            \"type\": \"Succeeded\",\n            \"status\": \"True\",\n            \"lastTransitionTime\": null,\n            \"reason\": \"Succeeded\"\n          }\n        ],\n        \"extraFields\": null\n      }\n    }\n  }\n"
```

Succeeded:
```
    logger.go:130: 2022-08-11T12:09:11.940-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:11.940-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T12:09:11.940-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T12:09:11.952-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:11.952-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded True  {0001-01-01 00:00:00 +0000 UTC} Succeeded }
    logger.go:130: 2022-08-11T12:09:11.952-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.successful.v1"
    logger.go:130: 2022-08-11T12:09:11.952-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_true_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
          specversion: 1.0
          type: dev.tekton.event.run.successful.v1
          source: /apis///namespaces/foo//test-run
          subject: test-run
          id: a1ada847-9b77-4c9c-bdf5-2b89c8cc9798
          datacontenttype: application/json
        Data,
          {
            "run": {
              "metadata": {
                "name": "test-run",
                "namespace": "foo",
                "resourceVersion": "00001",
                "creationTimestamp": null
              },
              "spec": {
                "serviceAccountName": ""
              },
              "status": {
                "conditions": [
                  {
                    "type": "Succeeded",
                    "status": "True",
                    "lastTransitionTime": null,
                    "reason": "Succeeded"
                  }
                ],
                "extraFields": null
              }
            }
          }
         already sent
=== RUN   TestReconcile_CloudEvents/Run_with_finished_false_condition
    logger.go:130: 2022-08-11T12:09:11.962-0400	INFO	cache/cacheclient.go:42	CACHE CLIENT &{lru:0xc0005f9320 lock:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0}}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-artifact-bucket" config was added or updated: &config.ArtifactBucket{Location:"", ServiceAccountSecretName:"", ServiceAccountSecretKey:"", ServiceAccountFieldName:"GOOGLE_APPLICATION_CREDENTIALS"}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-artifact-pvc" config was added or updated: &config.ArtifactPVC{Size:"5Gi", StorageClassName:""}
    logger.go:130: 2022-08-11T12:09:11.965-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-defaults" config was added or updated: &config.Defaults{DefaultTimeoutMinutes:60, DefaultServiceAccount:"default", DefaultManagedByLabelValue:"tekton-pipelines", DefaultPodTemplate:(*pod.Template)(nil), DefaultAAPodTemplate:(*pod.AffinityAssistantTemplate)(nil), DefaultCloudEventsSink:"http://synk:8080", DefaultTaskRunWorkspaceBinding:"", DefaultMaxMatrixCombinationsCount:256}
    logger.go:130: 2022-08-11T12:09:11.966-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "config-observability" config was added or updated: &config.Metrics{TaskrunLevel:"task", PipelinerunLevel:"pipeline", DurationTaskrunType:"histogram", DurationPipelinerunType:"histogram"}
    logger.go:130: 2022-08-11T12:09:11.966-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition.config-store	configmap/store.go:155	defaults/features/artifacts config "feature-flags" config was added or updated: &config.FeatureFlags{DisableAffinityAssistant:false, DisableCredsInit:false, RunningInEnvWithInjectedSidecars:true, RequireGitSSHSecretKnownHosts:false, EnableTektonOCIBundles:false, EnableCustomTasks:false, ScopeWhenExpressionsToTask:false, EnableAPIFields:"stable", SendCloudEventsForRuns:true, AwaitSidecarReadiness:true, EmbeddedStatus:"full"}
    logger.go:130: 2022-08-11T12:09:12.065-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:12.065-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded False  {0001-01-01 00:00:00 +0000 UTC} Cancelled }
    logger.go:130: 2022-08-11T12:09:12.065-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.failed.v1"
    logger.go:130: 2022-08-11T12:09:12.076-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:55	Reconciling test-run
    logger.go:130: 2022-08-11T12:09:12.076-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	run/run.go:67	Emitting cloudevent for test-run, condition: &{Succeeded False  {0001-01-01 00:00:00 +0000 UTC} Cancelled }
    logger.go:130: 2022-08-11T12:09:12.076-0400	DEBUG	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:149	Sending cloudevent of type "dev.tekton.event.run.failed.v1"
    logger.go:130: 2022-08-11T12:09:12.077-0400	INFO	TestReconcile_CloudEvents/Run_with_finished_false_condition	cloudevent/cloud_event_controller.go:157	cloudevent Context Attributes,
          specversion: 1.0
          type: dev.tekton.event.run.failed.v1
          source: /apis///namespaces/foo//test-run
          subject: test-run
          id: 12f13e3d-f297-41d0-a15a-1014a7768f1a
          datacontenttype: application/json
        Data,
          {
            "run": {
              "metadata": {
                "name": "test-run",
                "namespace": "foo",
                "resourceVersion": "00001",
                "creationTimestamp": null
              },
              "spec": {
                "serviceAccountName": ""
              },
              "status": {
                "conditions": [
                  {
                    "type": "Succeeded",
                    "status": "False",
                    "lastTransitionTime": null,
                    "reason": "Cancelled"
                  }
                ],
                "extraFields": null
              }
            }
          }
         already sent
```

Ok! I finally threw enough debug logging into the code to get something really useful:

```
    logger.go:130: 2022-08-12T14:08:00.771Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-12T14:08:00.772Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition run/run.go:67   Emitting cloudevent for test-run, condition: %!s(*apis.Condi
tion=<nil>)
    logger.go:130: 2022-08-12T14:08:01.299Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "
dev.tekton.event.run.started.v1"
CHECKING CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
EXISTING KEYS: []
    logger.go:130: 2022-08-12T14:08:01.381Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition run/run.go:55   Reconciling test-run
    logger.go:130: 2022-08-12T14:08:01.382Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition run/run.go:67   Emitting cloudevent for test-run, condition: %!s(*apis.Condi
tion=<nil>)
    logger.go:130: 2022-08-12T14:08:01.384Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:149        Sending cloudevent of type "
dev.tekton.event.run.started.v1"
CHECKING CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
EXISTING KEYS: []
    logger.go:130: 2022-08-12T14:08:01.378Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:169        past sending event dev.tekto
n.event.run.started.v1
    logger.go:130: 2022-08-12T14:08:01.386Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:172        Caching sent cloudevent dev.
tekton.event.run.started.v1
ABOUT TO CACHE ATTEMPT
ABOUT TO EVENTKEY
    logger.go:130: 2022-08-12T14:08:01.386Z     INFO    TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:169        past sending event dev.tekto
n.event.run.started.v1
GOT NEW CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
    logger.go:130: 2022-08-12T14:08:01.387Z     DEBUG   TestReconcile_CloudEvents/Run_with_no_condition cloudevent/cloud_event_controller.go:172        Caching sent cloudevent dev.
tekton.event.run.started.v1
ABOUT TO CACHE ATTEMPT
ABOUT TO EVENTKEY
NEW KEYS: [dev.tekton.event.run.started.v1/run/foo/test-run]
GOT NEW CACHE KEY dev.tekton.event.run.started.v1/run/foo/test-run
NEW KEYS: [dev.tekton.event.run.started.v1/run/foo/test-run]
    run_test.go:255: error in test Run with no condition: extra event received: "Context Attributes,\n  specversion: 1.0\n  type: dev.tekton.event.run.started.v1\n  source: /apis//
/namespaces/foo//test-run\n  subject: test-run\n  id: dfdea5ad-d200-40b7-9758-e528392f2ab1\n  datacontenttype: application/json\nData,\n  {\n    \"run\": {\n      \"metadata\": {\n
        \"name\": \"test-run\",\n        \"namespace\": \"foo\",\n        \"resourceVersion\": \"00001\",\n        \"creationTimestamp\": null\n      },\n      \"spec\": {\n
 \"serviceAccountName\": \"\"\n      },\n      \"status\": {\n        \"extraFields\": null\n      }\n    }\n  }\n"
```

What this means is that there are two calls to `SendCloudEventWithRetries` going on at the same time - you can tell, because `ABOUT TO CACHE ATTEMPT` should be showing up after `CHECKING CACHE KEY`, not interwoven as they are here. So we end up sending the event twice because it hasn't yet been added to the cache when the second call gets underway!

--- a/pkg/reconciler/events/cache/cache.go
+++ b/pkg/reconciler/events/cache/cache.go
@@ -31,21 +31,8 @@ type eventData struct {
 	Run *v1alpha1.Run `json:"run,omitempty"`
 }
 
-// AddEventSentToCache adds the particular object to cache marking it as sent
-func AddEventSentToCache(cacheClient *lru.Cache, event *cloudevents.Event) error {
-	if cacheClient == nil {
-		return errors.New("cache client is nil")
-	}
-	eventKey, err := EventKey(event)
-	if err != nil {
-		return err
-	}
-	cacheClient.Add(eventKey, nil)
-	return nil
-}
-
-// IsCloudEventSent checks if the event exists in the cache
-func IsCloudEventSent(cacheClient *lru.Cache, event *cloudevents.Event) (bool, error) {
+// ContainsOrAddCloudEvent checks if the event exists in the cache
+func ContainsOrAddCloudEvent(cacheClient *lru.Cache, event *cloudevents.Event) (bool, error) {
 	if cacheClient == nil {
 		return false, errors.New("cache client is nil")
 	}
@@ -53,7 +40,8 @@ func IsCloudEventSent(cacheClient *lru.Cache, event *cloudevents.Event) (bool, e
 	if err != nil {
 		return false, err
 	}
-	return cacheClient.Contains(eventKey), nil
+	isPresent, _ := cacheClient.ContainsOrAdd(eventKey, nil)
+	return isPresent, nil
 }
 
 // EventKey defines whether an event is considered different from another

--- a/pkg/reconciler/events/cache/cache_test.go
+++ b/pkg/reconciler/events/cache/cache_test.go
@@ -146,8 +146,8 @@ func TestAddCheckEvent(t *testing.T) {
 	for _, tc := range testcases {
 		t.Run(tc.name, func(t *testing.T) {
 			testCache, _ := lru.New(10)
-			AddEventSentToCache(testCache, tc.firstEvent)
-			found, _ := IsCloudEventSent(testCache, tc.secondEvent)
+			_, _ = ContainsOrAddCloudEvent(testCache, tc.firstEvent)
+			found, _ := ContainsOrAddCloudEvent(testCache, tc.secondEvent)
 			if d := cmp.Diff(tc.wantFound, found); d != "" {
 				t.Errorf("Cache check failure %s", diff.PrintWantGot(d))
 			}

--- a/pkg/reconciler/events/cloudevent/cloud_event_controller.go
+++ b/pkg/reconciler/events/cloudevent/cloud_event_controller.go
@@ -149,7 +149,7 @@ func SendCloudEventWithRetries(ctx context.Context, object runtime.Object) error
 		logger.Debugf("Sending cloudevent of type %q", event.Type())
 		// In case of Run event, check cache if cloudevent is already sent
 		if isRun {
-			cloudEventSent, err := cache.IsCloudEventSent(cacheClient, event)
+			cloudEventSent, err := cache.ContainsOrAddCloudEvent(cacheClient, event)
 			if err != nil {
 				logger.Errorf("error while checking cache: %s", err)
 			}
@@ -163,14 +163,9 @@ func SendCloudEventWithRetries(ctx context.Context, object runtime.Object) error
 			recorder := controller.GetEventRecorder(ctx)
 			if recorder == nil {
 				logger.Warnf("No recorder in context, cannot emit error event")
+				return
 			}
 			recorder.Event(object, corev1.EventTypeWarning, "Cloud Event Failure", result.Error())
-		}
-		// In case of Run event, add to the cache to avoid duplicate events
-		if isRun {
-			if err := cache.AddEventSentToCache(cacheClient, event); err != nil {
-				logger.Errorf("error while adding sent event to cache: %s", err)
-			}
 		}
 	}()
 

--- a/test/events/events.go
+++ b/test/events/events.go
@@ -26,6 +26,8 @@ import (
 // in the same order.
 func CheckEventsOrdered(t *testing.T, eventChan chan string, testName string, wantEvents []string) error {
 	t.Helper()
+	// Sleep 50ms to make sure events have delivered
+	time.Sleep(50 * time.Millisecond)
 	err := eventsFromChannel(eventChan, wantEvents)
 	if err != nil {
 		return fmt.Errorf("error in test %s: %v", testName, err)
@@ -37,6 +39,8 @@ func CheckEventsOrdered(t *testing.T, eventChan chan string, testName string, wa
 // were received via the given chan, in any order.
 func CheckEventsUnordered(t *testing.T, eventChan chan string, testName string, wantEvents []string) error {
 	t.Helper()
+	// Sleep 50ms to make sure events have delivered
+	time.Sleep(50 * time.Millisecond)
 	err := eventsFromChannelUnordered(eventChan, wantEvents)
 	if err != nil {
 		return fmt.Errorf("error in test %s: %v", testName, err)

